lib/src/rules/sort_public_methods_before_private_methods.dart

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) 2022, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+
+import '../analyzer.dart';
+
+const _desc = r'Define public methods before private methods.';
+
+const _details = r'''
+**DO** define public methods before private methods.
+
+This makes the code uniform across multiple classes
+and it makes it faster to find specific methods in a class. 
+
+**BAD:**
+```dart
+class A {
+  int a() => 0;
+  int _b() => 0;
+  int c() => 0;
+}
+```
+
+**GOOD:**
+```dart
+class A {
+  int a() => 0;
+  int c() => 0;
+  int _b() => 0;
+}
+```
+
+''';
+
+class SortPublicMethodsBeforePrivateMethods extends LintRule {
+  SortPublicMethodsBeforePrivateMethods()
+      : super(
+            name: 'sort_public_methods_before_private_methods',
+            description: _desc,
+            details: _details,
+            group: Group.style);
+
+  @override
+  void registerNodeProcessors(
+      NodeLintRegistry registry, LinterContext context) {
+    var visitor = _Visitor(this);
+    registry.addClassDeclaration(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+
+  _Visitor(this.rule);
+
+  void check(NodeList<ClassMember> members) {
+    bool privateMethodIsBeforePublicMethod = false;
+    // Members are sorted by source position in the AST.
+    for (var member in members) {
+      if (member is MethodDeclaration) {
+        if (privateMethodIsBeforePublicMethod &&
+            !_isPrivateName(member: member)) {
+          rule.reportLint(member.returnType);
```

Comment:
Do we want to report this lint for _every_ public member after a private member, or just the first?

Personally, I suspect that it's sufficient to report it for a single member.

Do we want to report this lint for public members or for private members? (That is, which does this lint consider to be in the wrong place?) The original name implies that the private members are in the right place and that it's the public members that need to be moved. The alternative I suggested (which might not be right) implies that the public members are right and that it's the private members that need to be moved.

Reply:
As far as I know other rules report a lint suggesting for every occurrence.
E.g. the rule `require_trailing_commas` tells me all the places where a comma is missing.
So for being consistent over multiple rules, I would change it to display all the occurrences simultaneously.
Also e.g. in Android Studio/IntelliJ the shortcut that moves the cursor to the next error would be faster if all the occurrences are mentioned at the same time, in comparison to requiring the analyzer to run again and then mention the next occurrence.

So if you are fine with that, I would try to implement to show all the occurrences at simultaneously.

Reply:
I'm open for suggestions if this lint should report for private or public members.
Personally I would not prefer one over the other, because when the lint is resolved, the result is the same.
What would you suggest?

Reply:
Regarding the report of multiple occurrences at one, this should already work.
I added a test to confirm that `test_multiple_cases_are_reported_simultaneously`.

Reply:
> As far as I know other rules report a lint suggesting for every occurrence.

Not quite, though I can see why it would seem that way.

The criteria we use is that every issue should have a single diagnostic. If there's a missing comma in one invocation, that's one issue and should have one diagnostic. If there's a missing comma in a second invocation, then that's a separate issue and should also have a diagnostic. Two separate issues, so two diagnostics.

But in this case it seems to me that the fact that the methods aren't sorted correctly is a single issue. For example, the code
```
class C {
  void m1() {}
  void _m2() {}
  void m3() {}
  void m4() {}
  void m5() {}
  void _m6() {}
}
```
Looks to me like someone added a private method (`m2`) in the wrong place. But if the lint reports that `m3`, `m4`, and `m5` are in the wrong place, that seems like noise to me. I'd much rather be told once that the methods aren't sorted correctly and then run "Sort Members" (or some other fix) to have it fix all the problems in this container.

It's possible that I'm the only one that feels that way.

> I'm open for suggestions if this lint should report for private or public members.

I don't know which is better. I believe that it depends on the failure mode that most users experience, which is data I don't have. I wondered whether you had an informed decision, and it sounds like you don't. It might not matter too much, but I tend to think about such things.

Reply:
I know understand what you mean and I think this is a better solution than linting every occurrence.
The linter should complain once per file to move the private methods below the public ones and a quick fix that does that would be a good solution.
I will try to implement that.

Reply:
A quick fix for that is not implemented in the linter package right?
I guess this is implemented in the dart sdk which cannot simply be extended by a merge request?
Or am I wrong?

Reply:
I'd hold off on doing more implementation work until we have gotten more feedback on the questions in the issue. We're being cautious about adding new stylistic lints and would like to gauge the reactions of the community before merging this PR.

> The linter should complain once per file to move the private methods ...

I suspect we'll end up with one lint per class, enum, etc.

> A quick fix for that is not implemented in the linter package right?
> I guess this is implemented in the dart sdk which cannot simply be extended by a merge request?

Yes, quick fixes are implemented in the `analysis_server` package. There's documentation for how to do that (https://github.com/dart-lang/sdk/blob/main/pkg/analysis_server/doc/implementation/quick_fix.md).

You can create a PR against the sdk repo and it will automatically converted into a Gerrit CL. If you look at the workflow steps in GitHub there should be a link to the CL. You won't have permissions to do much with the CL, but you can push new branches to update your PR and they'll also automatically get copied across.

But I'd also hold off on implementing a fix for a couple of reasons:
1. We already have a "Sort Members" function that could act as a fix (although it does more that might be expected from a fix for this lint). It isn't clear that we want to support yet another sorting feature.
2. Once this PR is merged (assuming it is) it can take a week or more for those changes to roll into the sdk, and you won't be able to test the fix until the lint is in the sdk.

Reply:
Alright, thanks for the information.

If the community likes this rule I would also try to create a PR in the sdk (assuming this one is merged).
I'm also looking forward for adding more rules if this is something the community would appreciate.

I would really like rules that sort members etc. like described in the [Flutter style guide](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#order-other-class-members-in-a-way-that-makes-sense).
So that in the end all of the mentioned rules for ordering can be automatically applied to files (if wished).

But I will wait some time and come back to this PR when more from the community have seen my proposal.
Thanks a lot for the help and feedback.